File: rubywarrior/howard-beginner/player.rb
class Player
  def initialize
    @health = 20
    @direction = :forward
  end

  def warrior_should_rest(warrior)
    # Check if health is low
    if warrior.health < 16
      # Health is low! Check if still taking damamge
      if @health == warrior.health
        # Not taking damage, rest!
        warrior.rest!
        @health = warrior.health + 2
      else
        # Still taking damage, if too low move away
        direction = @direction
        if warrior.health <= 10
          direction = @direction == :backward ? :forward : :backward
        end
        warrior.walk! direction
        @health = warrior.health
      end
    else
      warrior.walk! @direction
      @health = warrior.health if warrior.health == 10
    end
  end

  def look_around(warrior)
    if @health > warrior.health
      warrior.walk! @direction
      @health = warrior.health
      return true
    end
    spaces = warrior.look
    spaces.each do |space|
      return false if space.captive?
      if space.enemy?
        warrior.shoot! @direction
        return true
      end
    end
    false
  end

  def play_turn(warrior)
    if warrior.feel(@direction).empty?
      if warrior.health <= @health
        warrior_should_rest warrior unless look_around(warrior)
      else
        warrior_should_rest warrior
      end
    else
      warrior.pivot! if warrior.feel(@direction).wall?
      warrior.rescue! @direction if warrior.feel(@direction).captive?
      warrior.attack! @direction if warrior.feel(@direction).enemy?
      @health = warrior.health
    end
  end
end

File: warriorjs/george-beginner/Player.js
class Player {
    constructor() {
        this._health = 20;
        this._direction = 'forward';
    }

    // Check if the warrior should heal. Conditions for healing is if the warrior has less than a particular amount of
    // health and has not taken any damage recently. Otherwise the warrior will walk in the given direction.
    checkToHeal(warriorHealth) {
        if (warriorHealth < 18) {
            return this._health <= warriorHealth;
        }
        return false;
    }

    lookAround(warrior) {
        var spaces = warrior.look(this._direction);
        for (var i = 0; i < spaces.length; i++) {
            if (spaces[i].isCaptive()) {
                return false;
            }
            else if (spaces[i].isEnemy()) {
                warrior.shoot();
                return true;
            }
        }
        return false;
    }

    feelAround(warrior) {
        var space = warrior.feel(this._direction);
        if (space.isWall()) {
            warrior.pivot();
        } else if (space.isCaptive()) {
            warrior.rescue(this._direction);
        } else if (space.isEnemy()) {
            warrior.attack(this._direction);
        } else {
            warrior.walk(this._direction);
        }
    }

    playTurn(warrior) {
        if (this._health === warrior.health() && this.lookAround(warrior)) {

        }
        else if (warrior.feel(this._direction).isEmpty() && !warrior.feel(this._direction).isStairs()) {
            if (this.checkToHeal(warrior.health())) {
                warrior.rest();
            } else {
                if (this._health < 10) {
                    warrior.walk(this._direction === 'backward' ? 'forward' : 'backward');
                } else {
                    warrior.walk(this._direction);
                }
            }
        } else {
            this.feelAround(warrior);
        }
        this._health = warrior.health();
    }
}
